### ISSUE TITLE
possible issue with my_fdcopy() across file systems

### ISSUE THREAD

**Issue by stromsoe** [naemon/naemon-core#141]
I've noticed sending HUP or restarting naemon results in alerts being re-delivered and some times results in the loss of information that should be read from the state retention file (acknowledgement, comments, downtime).

I have the following variables set in naemon.cfg:

-----------------------------------------------
retain_state_information=1
use_retained_program_state=1
use_retained_scheduling_info=1
retention_update_interval=5
state_retention_file=/opt/ssd/naemon/var/retention.dat
temp_file=/opt/naemon/var/tmpfs/naemon.tmp
-----------------------------------------------

I monitor ~1800 hosts and ~135000 services.

Troubleshooting the problem shows that retention.dat is frequently truncated.  When the file is completely written out, it's roughly 162Mb.  The truncated files have varied anywhere from a few Mb up to 120Mb.

/opt/ssd is an SSD backed file system, and  /opt/naemon/var/tmpfs is a tmpfs backed filesystem.

xrddefault_save_state_information() writes out retention data to temp_file and calls my_rename() to rename temp_file to state_retention_file.    my_rename() fails on the rename and calls my_fcopy() which wraps my_fdcopy().

If I change temp_file to use the SSD backed file system instead of tmpfs, my_rename() uses the rename() syscall, and the complete state_retention_file gets written to disk correctly every time.

I have not started looking at my_fdcopy() to see why it's failing to complete writing out the file.  That's next on my list to do.  Any ideas?

**Comment by stromsoe** [comment]
Looking deeper, it looks like the bug may be in xrddefault_save_state_information().

My tmpfs is filling up and the temp file doesn't fit.  Return value for the fprintf() calls in xrddefault_save_state_information() isn't being checked, and the full filesystem condition is not being detected.

The temp file is closed normally (in the truncated state), and the truncated temp file is being copied over to the tmpfs file system.

**Comment by catharsis** [comment]
@stromsoe Good debugging. We should definitely be checking our returns, of course, though I'm not 100% sure what the correct action would be in case of an error. Two obvious options are:

* Refuse to restart if we can't write retention data (but then how do we communicate this to the user if the restart was initiated by a signal and not a synchronous command?)
* Restart, but ignore retaining status information altogether (at the cost of possible data loss)

In either case, we should be logging when this happens since it is an error that requires operator intervention.

Thoughts?

**Comment by stromsoe** [comment]
My preference would be to simplify the code and eliminate temp_file completely to remove the cross-file system problem.

Functions that write to temp_file and then call my_rename() could instead generate a temporary filename based on the destination filename.  my_rename() could then be simplified to  the standard rename() syscall, and my_fcopy() and my_fdcopy() could be removed.

That would allow xrddefault_save_state_information() and other functions that write to a temp file and rename it to abort the rename and log an error if there were any problems writing out to the temp file, leaving the "old" data file in place.

It's not optimal but it would preserve the last known state.

**Comment by catharsis** [comment]
Yup, that sounds like a sensible tradeoff. I'm not sure how soon I'll be able to get around to writing that patch myself, so if you'd be willing to submit a PR (assuming you're comfortable with that) that would be super helpful.

**Comment by stromsoe** [comment]
PR #142 created.  It modifies status.dat and retention.dat creation to use the values in status_file and retention_file as the input patterns to mkstemp() instead of temp_file, and modifies my_rename() to remove tests for errnor==XDEV and the file copy work-around and removes the now dead functions my_fcopy() and my_fdcopy().
